File: rewrite-core/src/main/java/org/openrewrite/marker/Markup.java
/*
 * Copyright 2022 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.marker;

import com.fasterxml.jackson.annotation.JsonCreator;
import lombok.EqualsAndHashCode;
import lombok.Value;
import lombok.With;
import org.openrewrite.Cursor;
import org.openrewrite.Incubating;
import org.openrewrite.RecipeScheduler;
import org.openrewrite.Tree;
import org.openrewrite.internal.ExceptionUtils;
import org.openrewrite.internal.RecipeRunException;
import org.openrewrite.internal.lang.Nullable;

import java.util.UUID;
import java.util.function.UnaryOperator;

import static org.openrewrite.Tree.randomId;

@Incubating(since = "7.31.0")
public interface Markup extends Marker {

    String getMessage();

    @Nullable
    String getDetail();

    @Override
    default String print(Cursor cursor, UnaryOperator<String> commentWrapper, boolean verbose) {
        if (verbose) {
            return commentWrapper.apply("(" + getDetail() + ")");
        }
        return commentWrapper.apply("(" + getMessage() + ")");
    }

    static <T extends Tree> T error(T t, Throwable throwable) {
        if (ExceptionUtils.containsCircularReferences(throwable)) {
            throwable = new Exception(throwable.getMessage());
            throwable.setStackTrace(throwable.getStackTrace());
        }
        return markup(t, new Markup.Error(randomId(), throwable));
    }

    static <T extends Tree> T warn(T t, Throwable throwable) {
        if (ExceptionUtils.containsCircularReferences(throwable)) {
            throwable = new Exception(throwable.getMessage());
            throwable.setStackTrace(throwable.getStackTrace());
        }
        return markup(t, new Markup.Warn(randomId(), throwable));
    }

    static <T extends Tree> T info(T t, String message) {
        return info(t, message, null);
    }

    static <T extends Tree> T info(T t, String message, @Nullable String detail) {
        return markup(t, new Markup.Info(randomId(), message, detail));
    }

    static <T extends Tree> T debug(T t, String message) {
        return debug(t, message, null);
    }

    static <T extends Tree> T debug(T t, String message, @Nullable String detail) {
        return markup(t, new Markup.Debug(randomId(), message, detail));
    }

    static <T extends Tree> T markup(T t, Markup markup) {
        return t.withMarkers(t.getMarkers().compute(markup, (s1, s2) -> s1 == null ? s2 : s1));
    }

    @Value
    @EqualsAndHashCode(onlyExplicitlyIncluded = true)
    @With
    class Error implements Markup {
        UUID id;

        @EqualsAndHashCode.Include
        String message;

        @EqualsAndHashCode.Include
        String detail;

        public Error(UUID id, Throwable exception) {
            this.id = id;
            Throwable cause = exception instanceof RecipeRunException ? exception.getCause() : exception;
            this.message = cause.getMessage();
            this.detail = ExceptionUtils.sanitizeStackTrace(cause, RecipeScheduler.class);
        }

        @JsonCreator
        public Error(UUID id, String message, String detail) {
            this.id = id;
            this.message = message;
            this.detail = detail;
        }
    }

    @Value
    @EqualsAndHashCode(onlyExplicitlyIncluded = true)
    @With
    class Warn implements Markup {
        UUID id;

        @EqualsAndHashCode.Include
        String message;

        @EqualsAndHashCode.Include
        String detail;

        public Warn(UUID id, Throwable exception) {
            this.id = id;
            Throwable cause = exception instanceof RecipeRunException ? exception.getCause() : exception;
            this.message = cause.getMessage();
            this.detail = ExceptionUtils.sanitizeStackTrace(cause, RecipeScheduler.class);
        }

        @JsonCreator
        public Warn(UUID id, String message, String detail) {
            this.id = id;
            this.message = message;
            this.detail = detail;
        }
    }

    @Value
    @EqualsAndHashCode(onlyExplicitlyIncluded = true)
    @With
    class Info implements Markup {
        UUID id;

        @EqualsAndHashCode.Include
        String message;

        @EqualsAndHashCode.Include
        @Nullable
        String detail;

        @Override
        public String getMessage() {
            return message;
        }

        @Nullable
        @Override
        public String getDetail() {
            return detail;
        }
    }

    @Value
    @EqualsAndHashCode(onlyExplicitlyIncluded = true)
    @With
    class Debug implements Markup {
        UUID id;

        @EqualsAndHashCode.Include
        String message;

        @EqualsAndHashCode.Include
        @Nullable
        String detail;

        @Override
        public String getMessage() {
            return message;
        }

        @Nullable
        @Override
        public String getDetail() {
            return detail;
        }
    }
}

File: rewrite-csharp/csharp/OpenRewrite/Core/Markup.cs
/*
 * Copyright 2026 the original author or authors.
 * <p>
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://docs.moderne.io/licensing/moderne-source-available-license
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
using OpenRewrite.Core.Rpc;
using OpenRewrite.Java;

namespace OpenRewrite.Core;

/// <summary>
/// Markers for annotating LST nodes with error, warning, info, or debug messages.
/// </summary>
public abstract class Markup(Guid id, string message, string? detail) : Marker, IEquatable<Markup>
{
    public Guid Id { get; } = id;
    public string Message { get; } = message;
    public string? Detail { get; } = detail;

    public bool Equals(Markup? other) => other is not null && Id == other.Id;
    public override bool Equals(object? obj) => Equals(obj as Markup);
    public override int GetHashCode() => Id.GetHashCode();

    private static T AddMarker<T>(T tree, Markup marker) where T : J
    {
        var newMarkers = tree.Markers.Add(marker);
        var withMarkers = tree.GetType().GetMethod("WithMarkers", [typeof(Markers)]);
        return withMarkers != null ? (T)withMarkers.Invoke(tree, [newMarkers])! : tree;
    }

    /// <summary>
    /// Adds an Error markup marker to the given tree node.
    /// </summary>
    public static T CreateError<T>(T tree, string message, string? detail = null) where T : J
        => AddMarker(tree, new Error(Guid.NewGuid(), message, detail));

    /// <summary>
    /// Adds a Warn markup marker to the given tree node.
    /// </summary>
    public static T CreateWarn<T>(T tree, string message, string? detail = null) where T : J
        => AddMarker(tree, new Warn(Guid.NewGuid(), message, detail));

    /// <summary>
    /// Adds an Info markup marker to the given tree node.
    /// </summary>
    public static T CreateInfo<T>(T tree, string message, string? detail = null) where T : J
        => AddMarker(tree, new Info(Guid.NewGuid(), message, detail));

    /// <summary>
    /// Adds a Debug markup marker to the given tree node.
    /// </summary>
    public static T CreateDebug<T>(T tree, string message, string? detail = null) where T : J
        => AddMarker(tree, new Debug(Guid.NewGuid(), message, detail));

    public sealed class Error(Guid id, string message, string? detail) : Markup(id, message, detail), IRpcCodec<Error>
    {
        public void RpcSend(Error after, RpcSendQueue q)
        {
            q.GetAndSend(after, m => m.Id);
            q.GetAndSend(after, m => m.Message);
            q.GetAndSend(after, m => m.Detail);
        }

        public Error RpcReceive(Error before, RpcReceiveQueue q)
        {
            var id = q.ReceiveAndGet<Guid, string>(before.Id, Guid.Parse);
            var message = q.Receive(before.Message);
            var detail = q.Receive(before.Detail);
            return new Error(id, message!, detail);
        }
    }

    public sealed class Warn(Guid id, string message, string? detail) : Markup(id, message, detail), IRpcCodec<Warn>
    {
        public void RpcSend(Warn after, RpcSendQueue q)
        {
            q.GetAndSend(after, m => m.Id);
            q.GetAndSend(after, m => m.Message);
            q.GetAndSend(after, m => m.Detail);
        }

        public Warn RpcReceive(Warn before, RpcReceiveQueue q)
        {
            var id = q.ReceiveAndGet<Guid, string>(before.Id, Guid.Parse);
            var message = q.Receive(before.Message);
            var detail = q.Receive(before.Detail);
            return new Warn(id, message!, detail);
        }
    }

    public sealed class Info(Guid id, string message, string? detail) : Markup(id, message, detail), IRpcCodec<Info>
    {
        public void RpcSend(Info after, RpcSendQueue q)
        {
            q.GetAndSend(after, m => m.Id);
            q.GetAndSend(after, m => m.Message);
            q.GetAndSend(after, m => m.Detail);
        }

        public Info RpcReceive(Info before, RpcReceiveQueue q)
        {
            var id = q.ReceiveAndGet<Guid, string>(before.Id, Guid.Parse);
            var message = q.Receive(before.Message);
            var detail = q.Receive(before.Detail);
            return new Info(id, message!, detail);
        }
    }

    public sealed class Debug(Guid id, string message, string? detail) : Markup(id, message, detail), IRpcCodec<Debug>
    {
        public void RpcSend(Debug after, RpcSendQueue q)
        {
            q.GetAndSend(after, m => m.Id);
            q.GetAndSend(after, m => m.Message);
            q.GetAndSend(after, m => m.Detail);
        }

        public Debug RpcReceive(Debug before, RpcReceiveQueue q)
        {
            var id = q.ReceiveAndGet<Guid, string>(before.Id, Guid.Parse);
            var message = q.Receive(before.Message);
            var detail = q.Receive(before.Detail);
            return new Debug(id, message!, detail);
        }
    }
}
